### PR TITLE
Jibsheet/member comparison

### PR DESCRIPTION
**Pull request by jibsheet**
I suggest reading both commits to see the before version we wrote and the after.

### CODE REVIEW

File: playbooks/library/mongodb_replica_set [hunk 1]
```diff
@@ -177,6 +177,38 @@ def fix_host_port(rs_config):
                 if 'port' in member:
                     del member['port']
 
+def compare_rs_config(old_config, new_config):
```
Comment (feanil) [inline playbooks/library/mongodb_replica_set]: I would maybe reword this to be something like `is_subset` instead of compare since it's not actually checking for equality.

File: playbooks/library/mongodb_replica_set [hunk 2]
```diff
@@ -177,6 +177,38 @@ def fix_host_port(rs_config):
...
+    '''
+    Compares the member list of a replica set configuration as specified (old_members)
+    to what Mongo has returned (new_members).  If it finds anything in old_members that
+    does not match new_members, it will return False.  new_members is allowed to contain
+    extra information that is not reflected in old_members because we do not necesarily
+    track all of mongo's internal data in the config.
+    '''
+    for member in old_members:
+        for k in member:
+            if member[k] != new_members[member['_id']][k]: return False
```
Comment (feanil) [inline playbooks/library/mongodb_replica_set:208]: nit: Put the `return False` on its own line.

**Comment by feanil** [comment]
Logic makes sense and is a lot easier to read than before.  The naming was misleading since it's not an equality check.

File: playbooks/library/mongodb_replica_set [hunk 3]
```diff
@@ -177,6 +177,38 @@ def fix_host_port(rs_config):
...
+
+    return True
+
```
Comment (maxrothman) [inline playbooks/library/mongodb_replica_set:211]: This entire function:

```
return any(v != new_members[member['_id'][k] for member in old_members for k,v in member.items())
```

File: playbooks/library/mongodb_replica_set [hunk 4]
```diff
@@ -226,11 +258,11 @@ def update_replset(rs_config):
 
     #Validate it worked
     if changed:
-        changed_rs_config = get_replset(client)
-        if changed_rs_config != rs_config:
-            module.fail_json(msg="Failed to validate that the replica set was changed", config=rs_config)
+        changed_rs_config = get_replset()
+        if not compare_rs_config(rs_config, changed_rs_config):
+            module.fail_json(msg="Failed to validate that the replica set was changed", changed_rs_config=changed_rs_config, config=rs_config)
```
Comment (maxrothman) [inline playbooks/library/mongodb_replica_set]: Since the existing results key is `config`, should `changed_rs_config` be `changed_config` instead?

Comment (jibsheet) [inline playbooks/library/mongodb_replica_set]: The existing results key is rs_config (The first argument to the function) so I don't understand the comment?

Ah - I see what you're commenting on again.  I didn't pick the internal variable names, and I don't see a ton of benefit in renaming them to strictly match what is returned from fail_json or exit_json a few lines later with the same data.

Comment (maxrothman) [inline playbooks/library/mongodb_replica_set]: No I'm suggesting changing this line to `module.fail_json(msg="...", changed_config=changed_rs_config, config=rs_config)` so that the user-facing names are consistent.

File: playbooks/library/mongodb_replica_set [hunk 5]
```diff
@@ -226,11 +258,11 @@ def update_replset(rs_config):
 
     #Validate it worked
     if changed:
-        changed_rs_config = get_replset(client)
-        if changed_rs_config != rs_config:
-            module.fail_json(msg="Failed to validate that the replica set was changed", config=rs_config)
+        changed_rs_config = get_replset()
+        if not check_config_subset(rs_config, changed_rs_config):
+            module.fail_json(msg="Failed to validate that the replica set was changed", changed_rs_config=changed_rs_config, config=rs_config)
```
Comment (maxrothman) [inline playbooks/library/mongodb_replica_set]: Porting old comment over:
I feel this line should be changed to `module.fail_json(msg="...", changed_config=changed_rs_config, config=rs_config)` to maintain consistency in user-facing names.

**Comment by maxrothman** [comment]
After comment addressed, :+1:
